File: Classes/GNApplicationController.m
//
//  GNApplicationController.m
//  Gmail Notifr
//
//  Created by James Chen on 1/27/13.
//  Copyright (c) 2013 ashchan.com. All rights reserved.
//

#import <MASShortcut/Shortcut.h>
#import "GNApplicationController.h"
#import "GNPreferences.h"
#import "GNAccount.h"
#import "GNBrowser.h"
#import "GNChecker.h"
#import "GNPreferencesController.h"
#import "GNAccountMenuController.h"

@interface GNApplicationController () <NSUserNotificationCenterDelegate>

@property (strong) IBOutlet NSMenu *menu;

@property (weak) IBOutlet NSMenuItem *menuItemCheckAll;
@property (weak) IBOutlet NSMenuItem *menuItemComposeMail;
@property (weak) IBOutlet NSMenuItem *menuItemPreferences;
@property (weak) IBOutlet NSMenuItem *menuItemAbout;
@property (weak) IBOutlet NSMenuItem *menuItemQuit;
@property (weak) IBOutlet NSMenuItem *menuItemRate;

@property (strong) NSStatusItem *statusItem;

@end

@implementation GNApplicationController {
    NSMutableArray *_checkers;
    NSMutableArray *_accountMenuControllers;
}

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
    self.statusItem = [[NSStatusBar systemStatusBar] statusItemWithLength:NSVariableStatusItemLength];
    [self.statusItem setHighlightMode:YES];
    [self.statusItem setMenu:self.menu];

    [self localizeMenuItems];

    self.statusItem.image = [NSImage imageNamed:@"NoMailsTemplate"];

    [GNPreferences setupDefaults];

    [self registerObservers];

    [self registerMailtoHandler];

    [self registerNotification];

    [self setupMenu];

    [self setupCheckers];

    [[MASShortcutBinder sharedBinder] bindShortcutWithDefaultsKey:GNDefaultsKeyCheckAllShortcut toAction:^{
        [self checkAll:nil];
    }];
}

- (void)applicationWillTerminate:(NSNotification *)notification {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (IBAction)showAbout:(id)sender {
    [[NSApplication sharedApplication] activateIgnoringOtherApps:YES];
    [[NSApplication sharedApplication] orderFrontStandardAboutPanel:sender];
}

- (IBAction)checkAll:(id)sender {
    [self checkAllAccounts];
}

- (IBAction)composeMail:(id)sender {
    GNAccount *account = [GNPreferences sharedInstance].accounts.firstObject;
    NSString *baseURL = account ? [account baseUrl] : @"https://mail.google.com/";
    NSString *url = [baseURL stringByAppendingString:@"?view=cm&tf=0&fs=1"];
    [self openURL:[NSURL URLWithString:url] withBrowserIdentifier:account.browser];
}

- (IBAction)showPreferencesWindow:(id)sender {
    [[NSApplication sharedApplication] activateIgnoringOtherApps:YES];
    [[GNPreferencesController sharedController] showWindow:sender];
}

- (IBAction)rateOnAppStore:(id)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"https://itunes.apple.com/app/gmail-notifr/id808154494?ls=1&mt=12"]];
}

- (void)showFAQs:(id)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"http://ashchan.com/projects/gmail-notifr#faqs"]];
}

- (void)showFeedback:(id)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"http://blog.ashchan.com/archive/2008/10/29/gmail-notifr-changelog/"]];
}

- (void)checkAccount:(id)sender {
    GNAccount *account = [self accountForGuid:[sender representedObject]];
    [[self checkerForAccount:account] reset];
}

- (void)openInbox:(id)sender {
    NSString *guid = [sender representedObject];
    [self openInboxForAccount:[self accountForGuid:guid]];

    // Check this account a short while after opening its inbox, so we don't have to check it
    // again manually just to clear the inbox count, since any unread mail is probably read now.
    // This can only be activated by a hidden default.
    NSTimeInterval autoCheckInterval = [GNPreferences sharedInstance].autoCheckAfterInboxInterval;
    if (autoCheckInterval > 0) {
        [[self checkerForGuid:guid] checkAfterInterval:autoCheckInterval];
    }
}

- (void)toggleAccount:(id)sender {
    GNAccount *account = [self accountForGuid:[sender representedObject]];
    account.enabled = !account.enabled;
    [account save];

    [self updateMenuItemAccountEnabled:account];
    [self updateMainMenu];
}

- (void)openMessage:(id)sender {
    [self openURL:[NSURL URLWithString:[sender representedObject]] withBrowserIdentifier:[GNAccount accountByMessageLink:[sender representedObject]].browser];
}

- (void)accountAdded:(NSNotification *)notification {
    if ([_accountMenuControllers count] == 1) {
        GNAccountMenuController *firstAccountMenuController = [_accountMenuControllers firstObject];
        [firstAccountMenuController detach];
        firstAccountMenuController.singleMode = NO;
        [firstAccountMenuController attachAtIndex:0 actionTarget:self];
        [[_checkers firstObject] reset];
    }

    GNAccount *account = [self accountForGuid:[notification userInfo][@"guid"]];
    [self createMenuForAccount:account atIndex:[[GNPreferences sharedInstance].accounts count] - 1];

    GNChecker *checker = [[GNChecker alloc] initWithAccount:account];
    [_checkers addObject:checker];
    [checker reset];

    [self updateMainMenu];
}

- (void)accountChanged:(NSNotification *)notification {
    GNAccount *account = [self accountForGuid:[notification userInfo][@"guid"]];
    [self updateMenuItemAccountEnabled:account];
    [[self checkerForAccount:account] reset];
}

- (void)accountRemoved:(NSNotification *)notification {
    GNAccountMenuController *menuController = [self menuControllerForGuid:[notification userInfo][@"guid"]];
    [menuController detach];
    [_accountMenuControllers removeObject:menuController];

    GNChecker *checker = [self checkerForGuid:[notification userInfo][@"guid"]];
    [checker cleanupAndQuit];
    [_checkers removeObject:checker];

    if ([_accountMenuControllers count] == 1) {
        GNAccountMenuController *singleAccountMenuController = [_accountMenuControllers firstObject];
        [singleAccountMenuController detach];
        singleAccountMenuController.singleMode = YES;
        [singleAccountMenuController attachAtIndex:0 actionTarget:self];
        [self checkAll:nil];
    } else {
        [self updateMenuBarCount:notification];
    }

    [self updateMainMenu];
}

- (void)accountsReordered:(NSNotification *)notification {
    NSMutableDictionary *menuControllers = [[NSMutableDictionary alloc] init];
    for (GNAccount *account in [GNPreferences sharedInstance].accounts) {
        GNAccountMenuController *controller = [self menuControllerForGuid:account.guid];
        [controller detach];
        menuControllers[account.guid] = controller;
    }

    for (NSUInteger i = 0; i < [[GNPreferences sharedInstance].accounts count]; i++) {
        GNAccount *account = [GNPreferences sharedInstance].accounts[i];
        [menuControllers[account.guid] attachAtIndex:i actionTarget:self];
    }

    [self checkAll:nil];
}

- (void)accountChecking:(NSNotification *)notification {
    [_statusItem setToolTip:NSLocalizedString(@"Checking Mail", nil)];
}

- (void)registerObservers {
    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
    [center addObserver:self selector:@selector(updateMenuBarCount:) name:GNShowUnreadCountChangedNotification object:nil];
    [center addObserver:self selector:@selector(accountAdded:) name:GNAccountAddedNotification object:nil];
    [center addObserver:self selector:@selector(accountChanged:) name:GNAccountChangedNotification object:nil];
    [center addObserver:self selector:@selector(accountRemoved:) name:GNAccountRemovedNotification object:nil];
    [center addObserver:self selector:@selector(updateAccountMenuItem:) name:GNAccountMenuUpdateNotification object:nil];
    [center addObserver:self selector:@selector(accountChecking:) name:GNCheckingAccountNotification object:nil];
    [center addObserver:self selector:@selector(accountsReordered:) name:GNAccountsReorderedNotification object:nil];
}

- (void)registerMailtoHandler {
    [[NSAppleEventManager sharedAppleEventManager] setEventHandler:self
                                                       andSelector:@selector(handleMailToEvent:eventReply:)
                                                     forEventClass:kInternetEventClass
                                                        andEventID:kAEGetURL];
}

- (void)handleMailToEvent:(NSAppleEventDescriptor *)event eventReply:(NSAppleEventDescriptor *)replyEvent {
    if ([[GNPreferences sharedInstance].accounts count] > 0) {
        GNAccount *account = [GNPreferences sharedInstance].accounts[0];
        NSString *link = [[event paramDescriptorForKeyword:keyDirectObject] stringValue];
        NSArray *urlComponents = [link componentsSeparatedByString:@"?"];
        NSString *recipients = [urlComponents[0] stringByReplacingOccurrencesOfString:@"mailto:" withString:@""];
        NSString *additionalParameters = @"";
        if ([urlComponents count] > 1) {
            // For some reason, Gmail does not interpret the query parameter "subject" correctly, and needs "su" instead.
            additionalParameters = [[NSString stringWithFormat:@"&%@",
                                     urlComponents[1]] stringByReplacingOccurrencesOfString:@"subject=" withString:@"su="];
        }
        NSString *url = [NSString stringWithFormat:@"%@?view=cm&tf=0&fs=1&to=%@%@",
                            [account baseUrl],
                            recipients,
                            additionalParameters];
        [self openURL:[NSURL URLWithString:url] withBrowserIdentifier:account.browser];
    }
}

- (void)registerNotification {
    [NSUserNotificationCenter defaultUserNotificationCenter].delegate = self;
}

- (void)userNotificationCenter:(NSUserNotificationCenter *)center didActivateNotification:(NSUserNotification *)notification {
    [self openInboxForAccountName:notification.title browser:[GNAccount accountByUsername:notification.title].browser];
    for (NSUserNotification *noti in center.deliveredNotifications) {
        if ([noti.title isEqualToString:notification.title]) {
            [center removeDeliveredNotification:noti];
        }
    }
}

- (void)setupMenu {
    _accountMenuControllers = [[NSMutableArray alloc] init];
    for (NSUInteger i = 0; i < [[GNPreferences sharedInstance].accounts count]; i++) {
        GNAccount *account = [GNPreferences sharedInstance].accounts[i];
        [self createMenuForAccount:account atIndex:i];
    }
}

- (void)setupCheckers {
    _checkers = [[NSMutableArray alloc] init];
    for (GNAccount *account in [GNPreferences sharedInstance].accounts) {
        [_checkers addObject:[[GNChecker alloc] initWithAccount:account]];
    }

    [self checkAllAccounts];
}

- (void)checkAllAccounts {
    for (GNChecker *checker in _checkers) {
        [checker reset];
    }
}

- (GNAccount *)accountForGuid:(NSString *)guid {
    for (GNAccount *account in [GNPreferences sharedInstance].accounts) {
        if ([account.guid isEqualToString:guid]) {
            return account;
        }
    }

    return nil;
}

- (GNChecker *)checkerForAccount:(GNAccount *)account {
    for (GNChecker *checker in _checkers) {
        if ([checker isForAccount:account]) {
            return checker;
        }
    }

    return nil;
}

- (GNChecker *)checkerForGuid:(NSString *)guid {
    for (GNChecker *checker in _checkers) {
        if ([checker isForGuid:guid]) {
            return checker;
        }
    }

    return nil;
}

- (void)openInboxForAccountName:(NSString *)name browser:(NSString *)browser {
    NSString *browserIdentier = browser ? browser : GNBrowserDefaultIdentifier;
    [self openURL:[NSURL URLWithString:[GNAccount baseUrlForUsername:name]] withBrowserIdentifier:browserIdentier];
}

- (void)openInboxForAccount:(GNAccount *)account {
    [self openInboxForAccountName:account.username browser:account.browser];
}

- (void)openURL:(NSURL *)url withBrowserIdentifier:(NSString *)browserIdentifier {
    if ([GNBrowser isDefault:browserIdentifier]) {
        [[NSWorkspace sharedWorkspace] openURL:url];
    } else {
        [[NSWorkspace sharedWorkspace] openURLs:@[url]
                        withAppBundleIdentifier:browserIdentifier
                                        options:NSWorkspaceLaunchDefault
                 additionalEventParamDescriptor:nil
                              launchIdentifiers:nil];
    }
}

#pragma mark - Menu Manipulation

- (void)updateMainMenu {
    // Update Check All
    NSString *stringToLocalize = [[GNPreferences sharedInstance].accounts count] <= 1 ? @"Check" : @"Check All";
    [self.menuItemCheckAll setTitleWithMnemonic:NSLocalizedString(stringToLocalize, nil)];
    NSArray *enabledAccounts = [[GNPreferences sharedInstance].accounts filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary *bindings) {
        return [evaluatedObject enabled];
    }]];
    [self.menuItemCheckAll setEnabled:[enabledAccounts count] > 0];

    // Show disable icon if no accounts enabled
    if (floor(NSAppKitVersionNumber) > NSAppKitVersionNumber10_9) {
        self.statusItem.button.appearsDisabled = [GNPreferences sharedInstance].allAccountsDisabled;
    }
}

- (void)localizeMenuItems {
    [self updateMainMenu];
    [self.menuItemComposeMail setTitleWithMnemonic:NSLocalizedString(@"Compose Mail", nil)];
    [self.menuItemPreferences setTitleWithMnemonic:NSLocalizedString(@"Preferences...", nil)];
    [self.menuItemAbout setTitleWithMnemonic:NSLocalizedString(@"About Gmail Notifr", nil)];
    [self.menuItemQuit setTitleWithMnemonic:NSLocalizedString(@"Quit Gmail Notifr", nil)];
    [self.menuItemRate setTitleWithMnemonic:NSLocalizedString(@"Rate on App Store", nil)];
}

- (void)createMenuForAccount:(GNAccount *)account atIndex:(NSUInteger)index {
    GNAccountMenuController *menuController = [[GNAccountMenuController alloc] initWithStatusItem:self.statusItem GNAccount:account];
    menuController.singleMode = [[GNPreferences sharedInstance].accounts count] == 1;
    [_accountMenuControllers addObject:menuController];
    [menuController attachAtIndex:index actionTarget:self];
}

- (void)updateAccountMenuItem:(NSNotification *)notification {
    GNAccount *account = [self accountForGuid:[notification userInfo][@"guid"]];
    GNAccountMenuController *menuController = [self menuControllerForGuid:account.guid];
    [menuController updateWithChecker:[self checkerForAccount:account]];
    [self updateMenuBarCount:notification];
}

- (void)updateMenuItemAccountEnabled:(GNAccount *)account {
    GNAccountMenuController *controller = [self menuControllerForGuid:account.guid];
    [controller updateStatus];
}

- (void)updateMenuBarCount:(NSNotification *)notification {
    NSMutableString *menuText = [NSMutableString new];
    NSMutableString *tooltipText = [NSMutableString new];
    NSUInteger messageCount = 0;
    BOOL useSeparateUnreadCounts = [GNPreferences sharedInstance].useSeparateUnreadCounts;
    for (GNChecker *checker in _checkers) {
        // show separate unread mail counts for each account (ie: "0:2")
        if (useSeparateUnreadCounts) {
            if ([menuText length] > 0) {
                [menuText appendString:@":"];
            }
            [menuText appendFormat:@"%d", (int)[checker messageCount]];
        }
        // show separate unread mail count in tooltips
        if ([tooltipText length] > 0) {
            [tooltipText appendString:@"\n"];
        }
        [tooltipText appendFormat:@"%@ (%d)", [checker.account username], (int)[checker messageCount]];

        messageCount += [checker messageCount];
    }

    if (!useSeparateUnreadCounts) {
        [menuText appendFormat:@"%d", (int)messageCount];
    }

    if (messageCount > 0 && [GNPreferences sharedInstance].showUnreadCount) {
        [_statusItem setTitle:menuText];
    } else {                                  
        [_statusItem setTitle:@""];
    }

    if (messageCount > 0) {
        [self.statusItem setToolTip:tooltipText];

        if ([GNPreferences sharedInstance].useColorIcon) {
            self.statusItem.image = [NSImage imageNamed:@"HaveMails-red"];
        } else {
            self.statusItem.image = [NSImage imageNamed:@"HaveMailsTemplate"];
        }
    } else {
        [self.statusItem setToolTip:@""];
        self.statusItem.image = [NSImage imageNamed:@"NoMailsTemplate"];
    }
}

- (GNAccountMenuController *)menuControllerForGuid:(NSString *)guid {
    for (GNAccountMenuController *controller in _accountMenuControllers) {
        if ([controller.guid isEqualToString:guid]) {
            return controller;
        }
    }

    return nil;
}

@end

File: Source/GNApplicationController.swift
//
//  GNApplicationController.h
//  Gmail Notifr
//
//  Created by James Chen on 1/27/13.
//  Copyright (c) 2013 ashchan.com. All rights reserved.
//

import Cocoa
import SwiftUI
import MASShortcut
import AppAuth
import GTMAppAuth
import GoogleAPIClientForREST_Gmail // SWIFT_PACKAGE=1 required, see https://github.com/google/google-api-objectivec-client-for-rest/issues/400

//@main
class GNApplicationController: NSObject, NSApplicationDelegate {
    var window: NSWindow!
    @IBOutlet var menu: NSMenu!
    @IBOutlet weak var menuItemCheckAll: NSMenuItem!
    @IBOutlet weak var menuItemComposeMail: NSMenuItem!
    @IBOutlet weak var menuItemPreferences: NSMenuItem!
    @IBOutlet weak var menuItemAbout: NSMenuItem!
    @IBOutlet weak var menuItemQuit: NSMenuItem!

    var statusItem: NSStatusItem!
    var checkers = [GNChecker]()
    var accountMenuControllers = [GNAccountMenuController]()

    var currentAuthorizationFlow: OIDExternalUserAgentSession?
    var authorization: GTMAppAuthFetcherAuthorization?
    static let clientID = "270244963224-8viqhtgpdks3vk56ffhvnfn112u4h26k.apps.googleusercontent.com"
    static let redirectURL = "com.googleusercontent.apps.270244963224-8viqhtgpdks3vk56ffhvnfn112u4h26k:/oauthredirect"

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        statusItem.menu = menu
        statusItem.button!.image = NSImage(named: "NoMailsTemplate")
        statusItem.button!.imagePosition = .imageLeft

        localizeMenuItems()

        GNPreferences.setupDefaults()

        registerObservers()
        registerURLHandler()
        registerNotification()

        setupMenu()
        setupCheckers()

        let contentView = MainView(selection: .constant(nil))

        window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: 480, height: 300),
            styleMask: [.titled, .closable, .miniaturizable, .resizable, .fullSizeContentView],
            backing: .buffered,
            defer: false
        )
        window.center()
        window.setFrameAutosaveName("Main Window")
        window.contentView = NSHostingView(rootView: contentView)

        NSApp.setActivationPolicy(.regular)
        window.makeKeyAndOrderFront(nil)

        MASShortcutBinder.shared().bindShortcut(withDefaultsKey: GNDefaultsKeyCheckAllShortcut) {
            self.checkAll(nil)
        }
   }

    func applicationWillTerminate(_ aNotification: Notification) {
        NotificationCenter.default.removeObserver(self)
    }

    func applicationSupportsSecureRestorableState(_ app: NSApplication) -> Bool {
        return true
    }

    @IBAction func showAbout(_ sender: AnyObject) {
        NSApplication.shared.activate(ignoringOtherApps: true)
        NSApplication.shared.orderFrontStandardAboutPanel(sender)
    }

    @IBAction func checkAll(_ sender: AnyObject?) {
        checkAllAccounts()
    }

    @IBAction func composeMail(_ sender: AnyObject) {
        let account = GNPreferences.sharedInstance().accounts.firstObject as? GNAccount
        let baseURL = account?.baseUrl() ?? "https://mail.google.com/"
        let url = baseURL + "?view=cm&tf=0&fs=1"
        openURL(url: URL(string: url)!, browserIdentifier: account?.browser ?? "")
    }

    @IBAction func showPreferencesWindow(_ sender: AnyObject) {
        NSApplication.shared.activate(ignoringOtherApps: true)
        GNPreferencesController.shared()?.showWindow(sender)
    }

    @IBAction func showFAQs(_ sender: AnyObject) {
        NSWorkspace.shared.open(URL(string: "http://ashchan.com/projects/gmail-notifr#faqs")!)
    }

    @IBAction func showFeedback(_ sender: AnyObject) {
        NSWorkspace.shared.open(URL(string: "http://blog.ashchan.com/archive/2008/10/29/gmail-notifr-changelog/")!)
    }

    @IBAction func rateOnAppStore(_ sender: AnyObject) {
        NSWorkspace.shared.open(URL(string: "https://itunes.apple.com/app/gmail-notifr/id808154494?ls=1&mt=12")!)
    }

    @objc func checkAccount(_ sender: AnyObject) {
        if let account = accountForGuid(guid: sender.representedObject as! String) {
            checkerForAccount(account: account)?.reset()
        }
    }

    @objc func openInbox(_ sender: AnyObject) {
        let guid = sender.representedObject as! String
        openInboxForAccount(account: accountForGuid(guid: guid)!)

        // Check this account a short while after opening its inbox, so we don't have to check it
        // again manually just to clear the inbox count, since any unread mail is probably read now.
        // This can only be activated by a hidden default.
        let autoCheckInterval = GNPreferences.sharedInstance().autoCheckAfterInboxInterval
        if autoCheckInterval > 0 {
            checkerForGuid(guid: guid)?.check(afterInterval: Int(autoCheckInterval))
        }
    }

    @objc func toggleAccount(_ sender: AnyObject) {
        let account = accountForGuid(guid: sender.representedObject as! String)!
        account.enabled.toggle()
        account.save()

        updateMenuItemAccountEnabled(account: account)
        updateMenu()
    }

    @objc func openMessage(_ sender: AnyObject) {
        openURL(url: URL(string: sender.representedObject as! String)!, browserIdentifier: GNAccount(byMessageLink: sender.representedObject as? String).browser)
    }

    @objc func accountAdded(_ notification: NSNotification) {
        if accountMenuControllers.count == 1 {
            let firstAccountMenuController = accountMenuControllers[0]
            firstAccountMenuController.detach()
            firstAccountMenuController.singleMode = false
            firstAccountMenuController.attach(at: 0, actionTarget: self)
            checkers[0].reset()
        }

        let account = accountForGuid(guid: notification.userInfo!["guid"] as! String)!
        createMenuForAccount(account: account, index: GNPreferences.sharedInstance().accounts.count - 1)

        let checker = GNChecker(account: account)!
        checkers.append(checker)
        checker.reset()

        updateMenu()
    }

    @objc func accountChanged(_ notification: NSNotification) {
        if let account = accountForGuid(guid: notification.userInfo!["guid"] as! String) {
            updateMenuItemAccountEnabled(account: account)
            checkerForAccount(account: account)?.reset()
        }
    }

    @objc func accountRemoved(_ notification: NSNotification) {
        let guid = notification.userInfo!["guid"] as! String
        let menuController = menuController(for: guid)!
        menuController.detach()
        accountMenuControllers.removeAll { c in
            c.guid == menuController.guid
        }
        let checker = checkerForGuid(guid: guid)
        checker?.cleanupAndQuit()
        checkers.removeAll { c in
            c.is(forGuid: guid)
        }

        if accountMenuControllers.count == 1 {
            let singleAccountMenuController = accountMenuControllers[0]
            singleAccountMenuController.detach()
            singleAccountMenuController.singleMode = true
            singleAccountMenuController.attach(at: 0, actionTarget: self)
            checkAll(nil)
        } else {
            updateMenuBarCount(notification)
        }

        updateMenu()
    }

    @objc func accountsReordered(_ notification: NSNotification) {
        var menuControllers = [String: GNAccountMenuController]()
        let accounts = GNPreferences.sharedInstance().accounts as! [GNAccount]
        for account in accounts {
            let controller = menuController(for: account.guid)!
            controller.detach()
            menuControllers[account.guid] = controller
        }

        for (index, account) in accounts.enumerated() {
            menuControllers[account.guid]?.attach(at: index, actionTarget: self)
        }

        checkAll(nil)
    }

    @objc func accountChecking(_ notification: NSNotification) {
        statusItem.button!.toolTip = NSLocalizedString("Checking Mail", comment: "")
    }

    func registerObservers() {
        let center = NotificationCenter.default
        center.addObserver(self, selector: #selector(updateMenuBarCount(_:)), name: NSNotification.Name.GNShowUnreadCountChanged, object: nil)
        center.addObserver(self, selector: #selector(accountAdded(_:)), name: NSNotification.Name.GNAccountAdded, object: nil)
        center.addObserver(self, selector: #selector(accountChanged(_:)), name: NSNotification.Name.GNAccountChanged, object: nil)
        center.addObserver(self, selector: #selector(accountRemoved(_:)), name: NSNotification.Name.GNAccountRemoved, object: nil)
        center.addObserver(self, selector: #selector(updateAccountMenuItem(_:)), name: NSNotification.Name.GNAccountMenuUpdate, object: nil)
        center.addObserver(self, selector: #selector(accountChecking(_:)), name: NSNotification.Name.GNCheckingAccount, object: nil)
        center.addObserver(self, selector: #selector(accountsReordered(_:)), name: NSNotification.Name.GNAccountsReordered, object: nil)
    }

    func registerURLHandler() {
         NSAppleEventManager.shared().setEventHandler(
            self,
            andSelector: #selector(handleURL(event:reply:)),
            forEventClass: AEEventClass(kInternetEventClass),
            andEventID: AEEventID(kAEGetURL)
        )
    }

    @objc func handleURL(event: NSAppleEventDescriptor, reply: NSAppleEventDescriptor) {
        let url = event.paramDescriptor(forKeyword: keyDirectObject)!.stringValue!

        if url.starts(with: "mailto:") {
            if GNPreferences.sharedInstance().accounts.count > 0 {
                let account = GNPreferences.sharedInstance().accounts[0] as! GNAccount
                let urlComponents = url.split(separator: "?")
                let recipients = urlComponents[0].replacingOccurrences(of: "mailto:", with: "")
                var additionalParameters = ""
                if urlComponents.count > 1 {
                    // For some reason, Gmail does not interpret the query parameter "subject" correctly, and needs "su" instead.
                    additionalParameters = String(format: "&%@", String(urlComponents[1])).replacingOccurrences(of: "subject=", with: "su=")
                }
                let url = "\(account.baseUrl()!)?view=cm&tf=0&fs=1&to=\(recipients)\(additionalParameters)"
                openURL(url: URL(string: url)!, browserIdentifier: account.browser)
            }
        } else {
            // OAuth
            currentAuthorizationFlow?.resumeExternalUserAgentFlow(with: URL(string: url)!)
        }
    }

    func setupMenu() {
        for (index, account) in GNPreferences.sharedInstance().accounts.enumerated() {
            createMenuForAccount(account: account as! GNAccount, index: index)
        }
    }

    func setupCheckers() {
        for account in GNPreferences.sharedInstance().accounts {
            checkers.append(GNChecker(account: account as? GNAccount))
        }

        checkAllAccounts()
    }

    func checkAllAccounts() {
        checkers.forEach { c in
            c.reset()
        }
    }

    func accountForGuid(guid: String) -> GNAccount? {
        (GNPreferences.sharedInstance().accounts as! [GNAccount]).first { $0.guid == guid }
    }

    func checkerForAccount(account: GNAccount) -> GNChecker? {
        checkers.first { $0.is(for: account) }
    }

    func checkerForGuid(guid: String) -> GNChecker? {
        checkers.first { $0.is(forGuid: guid) }
    }

    func messageCount() -> UInt {
        checkers.map({ $0.messageCount() }).reduce(0, +)
    }

    func openInboxForAccount(name: String, browser: String?) {
        let browserIdentier =  browser ?? GNBrowserDefaultIdentifier
        openURL(url: URL(string: GNAccount.baseUrl(forUsername: name))!, browserIdentifier: browserIdentier)
    }

    func openInboxForAccount(account: GNAccount) {
        openInboxForAccount(name: account.username, browser: account.browser)
    }

    func openURL(url: URL, browserIdentifier: String) {
        if GNBrowser.isDefault(browserIdentifier) {
            NSWorkspace.shared.open(url)
        } else {
            NSWorkspace.shared.open(
                [url],
                withAppBundleIdentifier: browserIdentifier,
                options: [],
                additionalEventParamDescriptor: nil,
                launchIdentifiers: nil
            )
        }
    }
}

// MARK: - Menu Manipulation
extension GNApplicationController {
    func updateMenu() {
        // Update Check All
        let stringToLocalize = GNPreferences.sharedInstance()!.accounts.count <= 1 ? "Check" : "Check All"
        menuItemCheckAll.title = NSLocalizedString(stringToLocalize, comment: "")
        let enabledAccounts = GNPreferences.sharedInstance()!.accounts.filter { a in
            (a as! GNAccount).enabled
        }
        menuItemCheckAll.isEnabled = !enabledAccounts.isEmpty

        // Show disable icon if no accounts enabled
        if (NSAppKitVersion.current > NSAppKitVersion.macOS10_9) {
            statusItem.button!.appearsDisabled = GNPreferences.sharedInstance().allAccountsDisabled
        }
    }

    func localizeMenuItems() {
        updateMenu()
        menuItemComposeMail.title = NSLocalizedString("Compose Mail", comment: "")
        menuItemPreferences.title = NSLocalizedString("Preferences...", comment: "")
        menuItemAbout.title = NSLocalizedString("About Gmail Notifr", comment: "")
        menuItemQuit.title = NSLocalizedString("Quit Gmail Notifr", comment: "")
    }

    func createMenuForAccount(account: GNAccount, index: Int) {
        let menuController = GNAccountMenuController(statusItem: statusItem, gnAccount: account)!
        menuController.singleMode = GNPreferences.sharedInstance().accounts.count == 1
        accountMenuControllers.append(menuController)
        menuController.attach(at: index, actionTarget: self)
    }

    func updateMenuItemAccountEnabled(account: GNAccount) {
        let controller = menuController(for: account.guid)
        controller?.updateStatus()
    }

    @objc func updateAccountMenuItem(_ notification: NSNotification) {
        let account = accountForGuid(guid: notification.userInfo!["guid"] as! String)!
        let menuController = menuController(for: account.guid)
        menuController?.update(with: checkerForAccount(account: account))
        updateMenuBarCount(notification)
    }

    @objc func updateMenuBarCount(_ notification: NSNotification) {
        let messageCount = messageCount()

        if messageCount > 0 && GNPreferences.sharedInstance().showUnreadCount {
            statusItem.button!.title = "\(messageCount)"
        } else {
            statusItem.button!.title = ""
        }

        if messageCount > 0 {
            var toolTipFormat = messageCount == 1 ? NSLocalizedString("Unread Message", comment: "") : NSLocalizedString("Unread Messages", comment: "")
#warning ("This is duplication. See GNChecker#processResult")
            if NSLocale.current.languageCode == "ru" {
                let count = messageCount % 100
                if (count % 10 > 4) || (count % 10 == 0) || (count > 10 && count < 15) {
                    toolTipFormat = NSLocalizedString("Unread Messages", comment: "")
                } else if count % 10 == 1 {
                    toolTipFormat = NSLocalizedString("Unread Message", comment: "")
                } else {
                    toolTipFormat = NSLocalizedString("Unread Messages 2", comment: "")
                }
            }

            statusItem.button!.toolTip = String(format: toolTipFormat, messageCount)
            statusItem.button!.image = NSImage(named: "HaveMailsTemplate")
        } else {
            statusItem.button!.toolTip = ""
            statusItem.button!.image = NSImage(named: "NoMailsTemplate")
        }
    }

    func menuController(for guid: String) -> GNAccountMenuController? {
        accountMenuControllers.first { $0.guid == guid }
    }
}

extension GNApplicationController: NSUserNotificationCenterDelegate {
    func registerNotification() {
        NSUserNotificationCenter.default.delegate = self
    }

    func userNotificationCenter(_ center: NSUserNotificationCenter, didActivate notification: NSUserNotification) {
        openInboxForAccount(name: notification.title!, browser: GNAccount(byUsername: notification.title).browser)
        for noti in center.deliveredNotifications {
            if noti.title == notification.title {
                center.removeDeliveredNotification(noti)
            }
        }
    }
}

// MARK: - Google OAuth
extension GNApplicationController {
    func authorize() {
        let request = OIDAuthorizationRequest(
            configuration: GTMAppAuthFetcherAuthorization.configurationForGoogle(),
            clientId: Self.clientID,
            scopes: [OIDScopeEmail, "https://www.googleapis.com/auth/gmail.readonly"],
            redirectURL: URL(string: Self.redirectURL)!,
            responseType: OIDResponseTypeCode,
            additionalParameters: nil
        )
        currentAuthorizationFlow = OIDAuthState.authState(byPresenting: request) { state, error in
            if let state = state {
                self.authorization = GTMAppAuthFetcherAuthorization(authState: state)
                // TODO: this is for experiment only
           }
        }
    }
}
